inc/blocks/namespace.php

```diff
@@ -93,6 +95,47 @@ function on_save_post( int $post_ID, WP_Post $post, bool $update ) : void {
 		return;
 	}
 
+	update_create_xb( $xbs, $post->post_title, $post_ID, null, 'post' );
+
+}
+
+/**
+ * Synchronise any XBs on the widget page with a shadow post type.
+ *
+ * @param array $instance The array of current widget settings.
+ * @param array $new_instance The array of new widget settings.
+ * @param array $old_instance The array of old widget settings.
+ * @param WP_Widget $widget The current widget instance.
+ * @return array $instance The current widget settings.
+ */
+function on_widgets_save( $instance, $new_instance, $old_instance, $widget ) {
+
+	if ( ! isset( $instance['content'] ) ) {
+		return $instance;
+	}
+
+	// Scan for XBs in the post content.
+	$blocks = parse_blocks( $instance['content'] );
+	$xbs = find_xbs( $blocks );
+
+	// Update / Create XB's if the content contains XB's.
+	if ( ! empty( $xbs ) ) {
+		update_create_xb( $xbs, 'Widget', null, $widget->number, 'widget' );
+	}
+
+	return $instance;
+}
+
+/**
+ * Update or create new XB's
+ *
+ * @param array $xbs XB's on page or as widgets.
+ * @param string $default_title Default Title title for XB Post.
+ * @param int $post_ID Post ID if applicable in use case.
+ * @param int $item_id Item ID.
+ * @param string $context Meta Key for updating the post_meta.
+ */
+function update_create_xb( array $xbs, string $default_title = '', ?int $post_ID = null, ?int $item_id = null, string $context = 'post' ) {
```

Comment:
Minor thing but for function names it's fine to be a bit more wordy if it makes the purpose more obvious e.g. `update_or_create_xb_post()`.

This is sometimes known as an "upsert" too e.g. update if thing exists else insert thing, but that terminology isn't used anywhere in WP.